pkg/exprhelpers/debugger.go

```diff
@@ -106,62 +107,30 @@ func (o *OpOutput) String() string {
 	return ret + ""
 }
 
-func (erp ExprRuntimeDebug) extractCode(ip int, program *vm.Program, parts []string) string {
+func (erp ExprRuntimeDebug) extractCode(ip int, program *vm.Program) string {
+	locations := program.Locations()
+	src := string(program.Source())
 
-	//log.Tracef("# extracting code for ip %d [%s]", ip, parts[1])
-	if program.Locations[ip].Line == 0 { //it seems line is zero when it's not actual code (ie. op push at the beginning)
-		log.Tracef("zero location ?")
-		return ""
-	}
-	startLine := program.Locations[ip].Line
-	startColumn := program.Locations[ip].Column
-	lines := strings.Split(program.Source.Content(), "\n")
-
-	endCol := 0
-	endLine := 0
-
-	for i := ip + 1; i < len(program.Locations); i++ {
-		if program.Locations[i].Line > startLine || (program.Locations[i].Line == startLine && program.Locations[i].Column > startColumn) {
-			//we didn't had values yet and it's superior to current one, take it
-			if endLine == 0 && endCol == 0 {
-				endLine = program.Locations[i].Line
-				endCol = program.Locations[i].Column
-			}
-			//however, we are looking for the closest upper one
-			if program.Locations[i].Line < endLine || (program.Locations[i].Line == endLine && program.Locations[i].Column < endCol) {
-				endLine = program.Locations[i].Line
-				endCol = program.Locations[i].Column
-			}
+	currentInstruction := locations[ip]
 
-		}
-	}
-	//maybe it was the last instruction ?
-	if endCol == 0 && endLine == 0 {
-		endLine = len(lines)
-		endCol = len(lines[endLine-1])
-	}
-	code_snippet := ""
-	startLine -= 1 //line count starts at 1
-	endLine -= 1
-
-	for i := startLine; i <= endLine; i++ {
-		if i == startLine {
-			if startLine != endLine {
-				code_snippet += lines[i][startColumn:]
-				continue
+	var closest *file.Location
+
+	for i := ip + 1; i < len(locations); i++ {
+		if locations[i].From > currentInstruction.From {
+			if closest == nil || locations[i].From < closest.From {
+				closest = &locations[i]
 			}
-			code_snippet += lines[i][startColumn:endCol]
-			break
-		}
-		if i == endLine {
-			code_snippet += lines[i][:endCol]
-			break
 		}
-		code_snippet += lines[i]
 	}
 
-	log.Tracef("#code extract for ip %d [%s] -> '%s'", ip, parts[1], code_snippet)
-	return cleanTextForDebug(code_snippet)
+	var end int
```

Comment:
Can we set end to len(src) and `closest != nil` then set closest.From?

Reply:
remove if else to a single if

Reply:
The case where `closest` is nil is the less common one (it will only happen for the last instruction of the expression), so we would be running a "useless" len each time.